keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,229 @@
+first time, there is no StorageState object. The controller creates the
+StorageState object for CronJob, setting .status.storedVersionsHash to
+["Unknown"] and .status.storageVersionHash to batch/v2alpha1. It also creates a
+storageVersionMigration object to request a migration. When the migration
+completed, the migration triggering controller sets .status.storedVersionsHash
+to [batch/v2alpha1].
+
+The migration triggering controller crashes and has been offline for less than
+10 minutes. When it's back online, it sees the .status.lastHeartbeatTime is less
+than 10 minutes old, thus it considers the storageState still valid. The
+controller does nothing other than updating the lastHeartbeatTime.
+
+The cluster admin upgrades the cluster to 1.15. Let's assume the storage version
+of the cronjob is changed to batch/v2beta1. The migration triggering controller
+observes the change. It sets the storageState.status.storageVersionHash to
+batch/v2beta1 and storageState.status.storedVersionHash to [v2alpha1, v2beta1].
+It also creates a migration request for cronjobs to migrate to v2beta1.
+
+The cluster admin finds a critical issue in 1.15 and downgrades the cluster
+back to 1.14. The migration triggering controller sets the
+storageState.status.storageVersionHash to batch/v2alpha1 and recreates the
+storageVersionMigration object to migrate to v2alpha1.
+
+## Future work: HA clusters
+
+When the masters of an HA cluster undergoes rolling upgrade, the storage
+versions might be configured differently in different API servers. Storage
+migration shouldn't start until the rolling upgrade is done.
+
+Ideally, for an HA cluster undergoing rolling upgrade, the following should
+happen for each resource: 
+* the discovery document should expose the lack of consensus by listing all
+  storage version hashes supported by different API servers.
+* the storage migration triggering controller should
+  * add all listed storage version hashes to storageState.status.storedVersionHashes.
+  * set storageState.status.storageVersionHash to `unknown`.
+  * delete any in progress migration by deleting existing
+    storageVersionMigrations.
+
+After the API server rolling upgrade is done, the discovery document would
+expose the agreed storage version hashes, and the storage migration triggering
+controller will resume the work described in the [previous section][].
+
+[previous section]:storage-migration-triggering-controller
+
+How to develop a convergence mechanism for HA masters is beyond the scope of
+this KEP. We will implement this KEP for non-HA case first. Operators of HA
+clusters need to disable the triggering controller and start migration
+manually, e.g., deploying the [initializer][] manually.
+
+[initializer]:https://github.com/kubernetes-sigs/kube-storage-version-migrator/tree/master/cmd/initializer
+
+## Future work: persisted discovery document
+
+Today, the discovery document is held in apiserver's memory. Previous state is
```

Comment:
If apiservers had a way to coordinate a version change, we'd have a reasonable place to send notifications about the version change. Right now we don't really :/